### PR TITLE
Streaks implementation

### PR DESCRIPTION
**Pull request by nada-ezziti**
### Pull Request: Add Streak Feature to Track User Activity

_____
**Description:**

This PR introduces the **streak feature**, allowing users to track:
1. The number of consecutive days they have been connected to the application.
2. The highest streak score (not yet implemented in the UI but planned for future integration, possibly in the StatsScreen).

The following changes have been made:

- **Added functionality:**
  - `updateStreak` and `getStreak` functions in the `UserViewModel` to handle streak calculation and retrieval.

- **UI updates:**
  - Displayed the updated streak value in the `ProfileScreen`.
  - Adapted the text to handle singular cases (e.g., "1 day") in the `Utils` file.
  - Changed the streak type to `Long` for compatibility with Firestore's number representation.

- **Testing:**
  - Added unit tests for `updateStreak` and `getStreak` to ensure correct behavior.

_____
**Next Steps:**
- Display the streak on other screens, starting with the `SettingsScreen`.

_____
**Impact:**
- Enhances user engagement by tracking daily activity.
- Prepares the groundwork for additional features on the StatsScreen.

____
**Note:**

I added two lines in `SettingsScreen` to display toast messages when the user saves or cancels information. This change is unrelated to the primary purpose of this pull request, but since there are already several pull requests awaiting review, I preferred not to open a separate one for such a minor update.

___
Capture d’écran 2024-11-14 à 02 46 49

### CODE REVIEW

File: app/src/main/java/com/github/se/signify/model/user/User.kt [hunk 1]
```diff
@@ -8,5 +8,8 @@ data class User(
     val friendRequests: List<String> = emptyList(),
     val friends: List<String> = emptyList(),
     val ongoingChallenges: List<String> = emptyList(), // New field for ongoing challenges
-    val pastChallenges: List<String> = emptyList() // New field for past challenges
+    val pastChallenges: List<String> = emptyList(), // New field for past challenges
+    val lastLoginDate: String = "",
+    val currentStreak: Long,
+    val highestStreak: Long
```
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/model/user/User.kt:14]: We can give default value to keep coherence with other element defined in the class.

File: app/src/main/java/com/github/se/signify/ui/Utils.kt [hunk 2]
```diff
@@ -539,12 +539,12 @@ fun LearnedLetterList(lettersLearned: List<Char>) {
...
  *
- * @param days An int value for the number of days (streak).
+ * @param days A long value for the number of days (streak).
  * @param daysText A boolean value => true will display the string " days" after the number.
  */
 @Composable
-fun StreakCounter(days: Int, daysText: Boolean) {
-  val text = if (daysText) " days" else ""
+fun StreakCounter(days: Long, daysText: Boolean) {
+  val text = if (!daysText) "" else if (days == 1L) " day" else " days"
```
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/Utils.kt:547]: A change in the call of `StreakCounter()` should be address in `AccountInformation()`.
You change the logic with the negation and now it don't show correctly the days in `ProfileScreen`, `FriendsListScreen` and `MyStatsScreen`.

File: app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt
Comment (NBCCF025) [inline app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:1]: It is great to have toast to give information to the user about the saved or canceled change in the profile. Well done, even if it is not related to the main purpose of this PR.

**Review by NBCCF025 — CHANGES_REQUESTED** [review]
Overall, this PR is a valuable addition, laying the foundation for user engagement with the streak tracking feature. The data is well stored in firestore.
But there is a need for improvement as the day streak is not correctly displayed in the UI. I request changes for that before merging. If you don't have time to do it, I will try to do the changes myself.

profileScreen
friendsListScreen
myStatsScreen

File: app/src/main/java/com/github/se/signify/model/user/User.kt (hunk 1 above)
Comment (nada-ezziti) [reply app/src/main/java/com/github/se/signify/model/user/User.kt]: Thanks for pointing this out. It's done.

**Review by NBCCF025 — APPROVED** [review]
It is all good now, well done!

Edit: the CI didn't pass because of test in HomeScreen that should be removed as the element no longer exist.

**Comment by NBCCF025** [comment]
Now everything is alright. Good job.

**Pull request merged**